io.openems.edge.controller.ess.cycle/src/io/openems/edge/controller/ess/cycle/statemachine/ContinueWithChargeHandler.java

```diff
@@ -1,43 +1,50 @@
 package io.openems.edge.controller.ess.cycle.statemachine;
 
+import java.time.LocalDateTime;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
 import io.openems.edge.common.statemachine.StateHandler;
 import io.openems.edge.controller.ess.cycle.statemachine.StateMachine.State;
+import io.openems.edge.ess.api.PowerConstraint;
+import io.openems.edge.ess.power.api.Phase;
+import io.openems.edge.ess.power.api.Pwr;
+import io.openems.edge.ess.power.api.Relationship;
 
 public class ContinueWithChargeHandler extends StateHandler<State, Context> {
 
 	private final Logger log = LoggerFactory.getLogger(ContinueWithChargeHandler.class);
 
 	@Override
-	public State runAndGetNextState(Context context) throws IllegalArgumentException, OpenemsNamedException {
+	public State runAndGetNextState(Context context) throws OpenemsNamedException {
+		final var controller = context.getParent();
 		final var ess = context.ess;
+		final var config = context.config;
+
+		if (config.maxSoc() == 100 && context.allowedChargePower == 0) {
+			return context.changeToNextState(State.COMPLETED_CYCLE);
+		}
 
-		if (context.config.maxSoc() > 99) {
-			if (context.maxChargePower == 0) {
-				// Wait for hysteresis
-				if (context.waitForChangeState(State.CONTINUE_WITH_CHARGE, State.COMPLETED_CYCLE)) {
-					return State.COMPLETED_CYCLE;
-				}
-				return State.CONTINUE_WITH_CHARGE;
-			}
-		} else if (ess.getSoc().orElse(0) >= context.config.maxSoc()) {
-			// Wait for hysteresis
-			if (context.waitForChangeState(State.CONTINUE_WITH_CHARGE, State.COMPLETED_CYCLE)) {
-				return State.COMPLETED_CYCLE;
-			}
-			return State.CONTINUE_WITH_CHARGE;
+		if (ess.getSoc().get() >= config.maxSoc()) {
+			return context.changeToNextState(State.COMPLETED_CYCLE);
 		}
 
-		// get max charge/discharge power
-		var power = context.getChargePower();
-		context.logInfo(this.log, "CONTINUE CHARGE with [" + power + " W]" //
-				+ " Current Cycle [ " + context.getParent().getCompletedCycles() + "] " //
-				+ "out of " + context.config.totalCycleNumber() + "]");
-		context.ess.setActivePowerLessOrEquals(power);
+		var power = context.getAcPower(ess, config.hybridEssMode(), config.power());
+		PowerConstraint.apply(ess, controller.id(), //
+				Phase.ALL, Pwr.ACTIVE, Relationship.EQUALS, -power);
+
+		context.logInfo(this.log, "CONTINUE CHARGE with [" + -power + " W]" //
+				+ " Current Cycle [ " + controller.getCompletedCycles() + "] " //
+				+ "out of " + config.totalCycleNumber() + "]");
 
 		return State.CONTINUE_WITH_CHARGE;
 	}
+
+	@Override
+	protected void onExit(Context context) throws OpenemsNamedException {
+		final var controller = context.getParent();
+		controller.setLastStateChangeTime(LocalDateTime.now(context.componentManager.getClock()));
```

Comment:
you could extract this to a method in Context or maybe add event callbacks to the StateMachine the you would have to only set it once

Reply:
actually we have somethng to set the next state, we can place it directly there. you have right.